### DISCUSSION TITLE
How to start the backend server without using docker?

### DISCUSSION THREAD

**Discussion by BladeSun** [fastapi/full-stack-fastapi-template#1111] [Questions] [answered]
How can I start the backend server without using docker?

**Comment by achaosss** [comment]
I have the same question.

**Comment by auh83** [comment]
Everything works for me with Docker. Without Docker, I tryed with virtualenv but it didn't find the solution for the moment...

**Comment by machov** [comment]
you guys can run it using `uvicorn main:app --reload` @auh83  @achaosss  @BladeSun

**Reply by DanielGuarnizo** [reply]
in which directory we should run this command ?

**Reply by machov** [reply]
`cd backend` the backend folder root, where the fastapi backend lives

**Comment by machov** [comment]
I'm struggling to understand which command is intitializing the backend in production?  because the Dockerfile doesn't have this `uvicorn main:app --reload`

**Reply by carlos-dz** [reply]
The docker-compose.override runs command: /start-reload.sh, it most be already on the image from where the backend is build tiangolo/uvicorn-gunicorn-fastapi:python3.10

**Reply by machov** [reply]
the Dockerfile runs uvicorn main:app --reload when importing the Dockerfile from [FROM tiangolo/uvicorn-gunicorn-fastapi:python3.10](https://github.com/tiangolo/uvicorn-gunicorn-docker/blob/7ff55d0d05ab9516f69fbefa69dc651612fd45fe/docker-images/python3.9-slim.dockerfile#L28C1-L28C18)
with command CMD ["/start.sh"] is there!

however `[COPY ./app /app/app](https://github.com/tiangolo/full-stack-fastapi-template/blob/master/backend/Dockerfile#L28) takes place till the end here .

How is unicorn being executed without the app? does this mean CMD at the end of tiangolo/uvicorn-gunicorn-fastapi:python3.10 is somehow delayed?

I see that docker-compose.override.yml has
command: [/start-reload.sh](https://github.com/tiangolo/uvicorn-gunicorn-docker/blob/7ff55d0d05ab9516f69fbefa69dc651612fd45fe/docker-images/start-reload.sh)
which executes uvicorn

however I can't understand how does docker-compose.yml work in production

**Reply by carlos-dz** [reply]
The Dockerfile just have instructions tu build an image, copy the files, install necessary packages on the image...doesn't run or execute the container. To run a single container from an image you would use the command docker run [image]... but docker compose let you build multiple images from different Dockerfiles and then run the containers for each image.

If you run "docker compose up -d" and then "docker compose exec backend batch" you can cd and explore the files inside the container and your will find start-reload.sh which executes uvicorn once the backend container is running.

**Comment by gospodima** [comment]
I have made a similar setup for backend without using docker. async sqlalchemy + pydantic schemas (without sqlmodels) are used there though.
Take a look: [Async FastAPI SQLAlchemy Template](https://github.com/gospodima/async-fastapi-sqlalchemy-template).

**Reply by auh83** [reply]
Thank you @gospodima !

**Comment by YuriiMotov** [comment, marked as answer]
It's described here in docs: https://github.com/fastapi/full-stack-fastapi-template/blob/master/development.md#local-development
In docs it's assumed you have already [activated venv and installed dependencies](https://github.com/fastapi/full-stack-fastapi-template/blob/master/backend/README.md#general-workflow)

So, basically:
1. activated venv and installed dependencies (do it once you created project)
```
# assuming you are in the root directory of the project
cd backend
uv sync
source .venv/bin/activate
cd ..
```

2. Run backend containers, stop container with fastapi app, run fastapi app locally (on host machine)
```
# assuming you are in the root directory of the project
docker compose watch
docker compose stop backend
cd backend
fastapi dev app/main.py
```
